File: DFS/90. Subsets II/Solution.py
# https://leetcode.com/problems/subsets-ii/

# By jiapeng
# Runtime: 40 ms, faster than 70.25% of Python3 online submissions for Subsets II.
# Memory Usage: 14 MB, less than 52.85% of Python3 online submissions for Subsets II.

class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        for i in range(len(nums)+1):
            self.dfs(0, i, nums, ans, [], 0)
        return ans
        
    def dfs(self, d, l, nums, ans, cur, starter):
        """
        traverse all possible answers of l-length
        
        Args:
            d: current length of cur
            l: target length of cur
            nums: the given distinct integers
            ans: final answer we want
            cur: one subset of ans in length l
            starter: index of the element our traverse start with
        """
        if d == l:
            ans.append(cur)
            return
        for i in range(starter, len(nums)):
            if l-d > len(nums)-i:
                break
            if i!=starter and nums[i]==nums[i-1]:   # remove duplicate subsets when nums contains duplicates
                continue
            self.dfs(d+1, l, nums, ans, cur+[nums[i]], i+1)

File: DFS/90. Subsets II/Solution.java
// https://leetcode.com/problems/subsets-ii/

/**
By jiapeng
Runtime: 1 ms, faster than 99.95% of Java online submissions for Subsets II.
Memory Usage: 39.9 MB, less than 59.04% of Java online submissions for Subsets II.
**/

class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        for(int i=0; i<=nums.length; i++)
            dfs(0, i, nums, ans, new ArrayList(), 0);
        return ans;
    }
    
    private void dfs(int d, int l, int[] nums, List<List<Integer>> ans, List<Integer> cur, int starter){
        if(d == l){
            ans.add(new ArrayList(cur));
            return;
        }
        for(int i=starter; i<nums.length; i++){
            if(l-d > nums.length-i)
                break;
            if(i != starter && nums[i] == nums[i-1])
                continue;
            cur.add(nums[i]);
            dfs(d+1, l, nums, ans, cur, i+1);
            cur.remove(cur.size()-1);
        }
    }
}
